### ISSUE TITLE
CLI and Bearer Tokens

### ISSUE THREAD

**Issue by ambsw-technology** [docker/docker#29257]
I'm attempting to provide a bearer token to authenticate the CLI by adding a header to `config.json` and it isn't working.

I'm using Fabric to execute remote deployment scripts.  For several reasons -- including [these audit log benefits](http://www.projectatomic.io/blog/2015/08/why-we-dont-let-non-root-users-run-docker-in-centos-fedora-or-rhel/) -- I'd like to continue to use the CLI (instead of using the REST API).  However, we're using private repositories and I'd rather not pipe my Docker password to every single endpoint for `docker login` so I'd like to use Bearer tokens.  I found the same question, unanswered, [on the Docker forum](https://forums.docker.com/t/token-based-authentication/11742).

I believe I have worked out the "correct" approach.  Specifically, the documentation on the [Swarm API](https://docs.docker.com/swarm/swarm-api/#/authenticate-using-registry-tokens) includes the use of the `HttpHeaders` section of `~/.docker/config.json` to include a token in CLI calls.  In principle, the same strategy can be used to provide the (properly formatted) [Bearer token as an Authorization header](https://docs.docker.com/registry/spec/auth/jwt/#/using-the-signed-token).

Of course, I'm here because it isn't working.  My testing is well within the 300 second expiration window (and I realize I will need to continually refresh the token).  In all my reading, the only reason I can find to explain this issue is that "Docker does not allow these headers to change any headers it sets for itself." ([manpage](http://manpages.ubuntu.com/manpages/xenial/man5/config-json.5.html)).  If this is the reason, I guess I'm submitting a feature request to use config.json as the default (when not logged in) instead of a bug.

I'm running `Docker version 1.12.3, build 6b644ec`.  My config.json (token shortened for security/readability) is:

```
{
    "HttpHeaders" : {
        "Authorization" : "Bearer ey...hw"
    }
}
```

I'm happy to provide `docker info` and others if it's helpful, but I don't think this is that kind of problem.

**Comment by thaJeztah** [comment]
> I'm happy to provide docker info and others if it's helpful, but I don't think this is that kind of problem.

Are you currently running the official binaries, or the Red Hat version? Note that the "audit" logging you're referring to is only present in the Red Hat fork of docker, not in the official binaries (the patch wasn't accepted because of various technical issues with the code; https://github.com/docker/docker/pull/14446).


I'm trying to reproduce, but looks like it's working;

create configuration file;

```bash
mkdir -p ~/.docker
cat > ~/.docker/config.json <<EOF
{
    "HttpHeaders" : {
        "Authorization" : "Bearer foobar"
    }
}
EOF
```

just for debugging, listen on localhost:2375

```bash
nc -l -p 2375
```

In another shell, call a docker command;

```
docker -H localhost:2375 info
```

And see the output;

```
GET /v1.24/info HTTP/1.1
Host: localhost:2375
User-Agent: Docker-Client/1.12.3 (linux)
Authorization: Bearer foobar
Accept-Encoding: gzip
```


I see the article you're refering to uses a wrapper that calls `sudo` for the
docker commands. Depending on how you're using `sudo`, it may be worth
checking if it's looking in the "root's" homedir (e.g. `/root`) for the configuration file?

**Comment by ambsw-technology** [comment]
Thanks.  I'm using production binaries.  I'm actually trying to get both (1) LDAP control over docker execution and (2) audit logging of these commands.  The article gave me the general strategy to achieve both without needing special binaries:

 - Leave users out of the `docker` group
 - Create an LDAP group with passwordless access to `sudo docker`
 - Alias `sudo docker` to `docker` so it's transparent
 - Audit log `sudo`

And your tests seem to indicate that the issue isn't the header:

`docker -H localhost:2375 info`

```
GET /v1.24/info HTTP/1.1
Host: localhost:2375
User-Agent: Docker-Client/1.12.3 (linux)
Authorization: Bearer foobar
Accept-Encoding: gzip
```

But I get the right token -- with inline auth to eliminate all of the issues I can think of:

```
https://<username>:<password>@auth.docker.io/token?service=registry.docker.io&scope=repository:ambsw/nginx:pull
```

I change the Bearer token and retest to verify the token has changed:

```
GET /v1.24/info HTTP/1.1
Host: localhost:2375
User-Agent: Docker-Client/1.12.3 (linux)
Authorization: Bearer ey...5g
Accept-Encoding: gzip
```

Perfect.  Now I try to pull the repository:

```
$docker pull ambsw/nginx
Using default tag: latest
Pulling repository docker.io/ambsw/nginx
Error: image ambsw/nginx:latest not found
```

Just to make sure I didn't fat finger something...

```
$docker login
<same credentials as above>
$docker pull ambsw/nginx
Using default tag: latest
latest: Pulling from ambsw/nginx
Digest: sha256:67477cb767a2c0c1e6795d8b1fe2d332488e90964d1b594d852a00c58069097c
Status: Image is up to date for ambsw/nginx:latest
```

So you're right that the header seems to be passed by the CLI.  Is that not enough?

**Comment by ambsw-technology** [comment]
If we exclude the CLI, let's assume my problem is on the token side.  Is there an easy way to introspect a token to verify the permissions it carries?

**Comment by ambsw-technology** [comment]
Found the [JWT Debugger](https://jwt.io/).  I assume `access: []` is my problem (slightly sanitized):

```
PAYLOAD:

{
  "access": [],
  "aud": "registry.docker.io",
  "exp": 1481288289,
  "iat": 1481287989,
  "iss": "auth.docker.io",
  "jti": "<...>",
  "nbf": <...>,
  "sub": ""
}
```

Obviously the CLI doesn't give me the same 401 feedback to verify that my parameters are right, but the URL above aligns with every example I can find.

**Comment by thaJeztah** [comment]
If this is for authenticating to docker hub, would using an external credentials store help? https://docs.docker.com/engine/reference/commandline/login/#/credentials-store

**Comment by ambsw-technology** [comment]
> would using an external credentials store help? 

If you're asking me, I certainly wouldn't know. If you're recommending that I try, I'm happy to.  However, I should probably resolve my question on `access: []` before I get too far along.  If `access` is a problem, nothing I do to the CLI will matter.

**Comment by dmcgowan** [comment]
Docker Hub currently does not support [long lived refresh tokens](https://docs.docker.com/registry/spec/auth/oauth/#example-getting-refresh-token). The bearer tokens linked are short lived and only intended for an immediate action on a single repository.

The only solutions I could suggest with private repositories would be implement a credential store to deliver the credentials to the docker cli or distribute your config.json with the credentials. If the hub does turn on support for refresh tokens in the future then the method of distributing the config.json would still apply, just instead of the base64 encoded basic auth creds, it would have the refresh token.

**Comment by ambsw-technology** [comment]
I'm confused.  I'm talking about using Bearer tokens.  As part of my build script, I intended to repeatedly acquire and pass to the client the latest Bearer token.  I didn't intend to (and don't think I am) using long lived refresh tokens.

 - Are you saying that bearer tokens are not supported by Docker Hub?  
 - If bearer tokens are supported, can you confirm if the `access: []` section is OK or should really say something like `access: [pull]`? 
 - If the section should say `access: [pull]`, do you see an obvious issue in my token request?
 - If the token request looks good, what permissions need to exist on the Hub?  I'm the owner of this organization and (as I showed in testing) have no problems pulling using simple auth.

**Comment by ambsw-technology** [comment]
OK.  I confirmed that my access section does not match the [documentation](https://docs.docker.com/registry/spec/auth/jwt/):

```
...
    "access": [
        {
            "type": "repository",
            "name": "samalba/my-app",
            "actions": [
                "pull",
                "push"
            ]
        }
...
```

My request also seems to match the request in the documentation (I've repeated with `pull,push` with no change):

 - mine: `https://<username>:<password>@auth.docker.io/token?service=registry.docker.io&scope=repository:ambsw/nginx:pull`
 - docs; `https://auth.docker.io/token?service=registry.docker.io&scope=repository:samalba/my-app:pull,push`

However, I finally resolved the issue after stumbling upon [this script](https://gist.github.com/jlhawn/8f218e7c0b14c941c41f).  Apparently, something between Chrome and `auth.docker.io` doesn't like the `username:password@<url>` strategy.  Once I sent the basic authentication as a header (`Authorization: Basic <base64 encoded username:password>`) I got a token with the correct access section.

**Comment by ambsw-technology** [comment]
Correction.  I was able to resolve the `access` issue:

```
  "access": [
    {
      "type": "repository",
      "name": "ambsw/nginx",
      "actions": [
        "pull"
      ]
    }
  ],
  "aud": "registry.docker.io",
  "exp": 1482015423,
  "iat": 1482015123,
  "iss": "auth.docker.io",
  "jti": "d0xszb-b5gqnTS5lukyL",
  "nbf": 1482014823,
  "sub": "<sub>"
```

However, I'm still having issues on the CLI:

```
$ docker pull ambsw/nginx:latest
Pulling repository docker.io/ambsw/nginx
Error: image ambsw/nginx:latest not found
```

To be clear, I've created tokens with both `"aud": "registry.docker.io"` (matching the docs except for `.io` vs `.com` and `"aud": "docker.io"` (matches the default in the client).

**Comment by dmcgowan** [comment]
@ambsw-technology you should not be making any alterations to the bearer token, as it will fail the signature check. The only flow we support is getting the short lived bearer token from the token API and immediately using this token with the registry. All these interactions happen on the daemon and attempting to manage the bearer token on the client side is not supported. Docker login is definitely the easiest way, but either passing around a docker config (which was previously created with docker login) or using a credential manager are your best bets. Since the hub is not supporting refresh token right now, I would suggest trying to use the credential manager as the most secure option.

**Comment by ambsw-technology** [comment]
I was only sanitizing posts for Git... I use the unaltered tokens during the actual flow.

It seems like this is a quintessential use case for bearer tokens -- passing them to a deployment target in lieu of a barely obfuscated username/password.  Should I open a separate feature request or do you want to reopen this (with a robust history of the existing state) and flag it as such?

**Comment by dmcgowan** [comment]
@ambsw-technology we have already implemented this use case with long lived refresh tokens. If the registry/token server does not support it then yes you are forced to use username/password. Whether obfuscated or stored securely within a password manager using the credential store, I agree that it is still not an ideal setup. However the bearer tokens handed to the registry are defined to be short lived with limited scope and returned from the token server as such, that is why I don't consider supporting them through the config as a feature. Also consider that the client does not know the duration a push/pull operation will take nor the scopes the daemon will ask for.

**Comment by cirocosta** [comment]
Hey @dmcgowan , sorry if I'm being pedantic, but when I get the token from `https://auth.docker.io/token?scope=repository:$image:pull&service=registry.docker.io`, is it supposed to be used right away and not anymore, right? Is it possible to retrieve a long lived token from Docker Hub?

 Thx!

**Comment by dmcgowan** [comment]
@cirocosta the token returned is short lived. The hub currently does not implement the long lived token specification.

**Comment by vpaul** [comment]
I hit the same issue. The token is issued for 5 mins by default. It is enough time to pull/push. With some curl or a like scripting it is possible to pull by layers and feed to `docker load`. But adding httpheaders with the same authorization header as to curl to config.json for regular docker client doesn't make any sense.

**Comment by YakobovLior** [comment]
So the authorization header should only be used with API requests and not with the docker client?

**Comment by vpaul** [comment]
right. The token is short-live and can't be used in a permanent config. Your API code should include get token part. Thinking out loud you can try updating config.json periodically but never tested this.

**Comment by YakobovLior** [comment]
> right. The token is short-live and can't be used in a permanent config. Your API code should include get token part. Thinking out loud you can try updating config.json periodically but never tested this.

Thank you for the reply.
Actually the idea of updating the config.json rather than writing a pull mechanism with API requests sounds easier, but the problem is that I can't make the docker client pull images with the bearer token.
This is my config.json file: (trimmed for security considerations).
{
        "HttpHeaders": {
                "Authorization": "Bearer eyJhY...SJ9",
                "User-Agent": "Docker-Client/1.13.1 (linux)"
        }
}

The token has the correct data in access, but still docker client refuses to pull until I make a regular login.
Any ideas?

**Comment by ambsw-technology** [comment]
@YakobovLior, I aspired to the exact same behavior when I opened this ticket, but the conclusion is that an `Authorization` header is not supported in this use case.  I think the basic logic is that the JSON file is treated as a long-lived structure.  They don't think it's logical to embed a short-lived token in a long-lived structure.  

While I understand the perspective, the short-lived tokens are the **right** thing to pass around.  I want to give authorization to a deployment target to execute a single, narrow command.  For the added security (we're a Healthcare IT company) I'm happy (even eager) to send a new bearer token before/with command.  Storing a username/password isn't a proper substitute because it opens pandora's box.

The only "supported" way to use this security feature is to make an API call instead of using the CLI.  Unfortunately, the last time I looked, it's nearly impossible to log API calls.  So I'm in a catch-22.  Either I pass a username/password to the machine (bad) and log the `sudo docker` calls (we don't put any users in the `docker` group to force them to use `sudo`) or I make a call that I can't record in my security audit trails (worse).
